test/type-system.ts

```diff
@@ -294,3 +295,44 @@ test("it handles complex types correctly", t => {
 
     t.is(true, true) // supress no asserts warning
 })
+
+test("it should provide detailed reasons why the value is not appicable", t => {
+    const Todo = types.model({
+        title: types.string
+    }, {
+        setTitle(v: string) {
+
+        }
+    })
+
+    const Store = types.model({
+        todos: types.map(Todo),
+        get amount() {
+            // double check, not available design time:
+            /// this.setAmount()
+            return this.todos.size
+        },
+        getAmount(): number {
+            return this.todos.size + this.amount
+        }
+    }, {
+        setAmount() {
+            const x: number = this.todos.size + this.amount + this.getAmount
+        }
+    })
+
+    t.throws(() => Store.create({
+        todos: {
+            "1":{
+                title: true,
+                setTitle: "hello"
+            }
+        },
+        amount: 1,
+        getAmount: "hello"
+    }), `[mobx-state-tree] Error while converting {"todos":{"1":{"title":true,"setTitle":"hello"}},"amount":1,"getAmount":"hello"} to AnonymousModel:
```

Comment:
This is really awesome and impressive! Will update my slides :). Really awesome work!